Update fe forcing

mnlevy1981 opened the pull request:
https://github.com/marbl-ecosys/MARBL/pull/464 updates MARBL's iron sediment forcing scheme, which will change what forcing fields MOM6 needs to provide to MARBL

mnlevy1981 committed:
Use updated iron sediment forcing scheme

Instead of just fesedflux and feventflux, updates to MARBL from UCI now add an
fesedfluxred forcing field.

mnlevy1981 committed:
Merge tag 'dev/ncar_250128' into update_fe_forcing

mnlevy1981 committed:
Merge tag 'dev/ncar_250428' into update_fe_forcing

mnlevy1981 committed:
New MOM_initialize_tracer_from_Z() argument

Added optional argument num_pass to MOM_initialize_tracer_from_Z(), which then
gets passed down to horiz_interp_and_extrap_tracer() eventually is passed to
fill_miss_2d() [where the optional argument is already part of the API but is
never used]. I tested this updated by setting num_pass=0 for the MARBL tracers
and verifying that the smoothing step was not run, but that was just a test to
see if the smoothing was responsible for some unexpected tracer values in the
MARBL initial conditions -- in general, we do want to smooth the MARBL tracers
after they are initialized.

mnlevy1981 committed:
Use MARBL's autotroph_tracer_consistency_enforce()

There was an issue where smoothing the MARBL tracer initial conditions led to
some inconsistent tracer values for tracers associated with MARBL's autotrophs.
For each autotroph, if any tracer (C, N, P, etc) is 0 we want all of that
autotroph's tracers to be 0... but the smoothing algorithm in
MOM_initialize_tracer_from_Z() would sometimes leave one tracer at 0 but
introduce a nonzero value in a different tracer at the same location. The
subroutine on the MARBL interface enforces this consistency.

This function is called whenever any MARBL tracer is initialized with
MOM_initialize_tracer_from_Z(); if all the tracers are read in from a restart
file then this subroutine is skipped. If this function is called, it is noted
in the log with the following message:

Enforcing consistency across autotroph tracer initial conditions

mnlevy1981 committed:
Dummy MARBL interface missing subroutine

Added autotroph_tracer_consistency_enforce() to the dummy MARBL interface so
MARBL_tracers.F90 can still be built without MARBL

mnlevy1981 committed:
Remove unnecessary warning messages

MARBL_tracers_stock() was writing a list of all tracers from all processors,
filling up cesm.log unnecessarily

mnlevy1981 committed:
Merge branch 'dev/ncar' into update_fe_forcing

mnlevy1981 committed:
Merge tag 'dev/ncar_251024b' into update_fe_forcing

mnlevy1981 committed:
Merge tag 'dev/ncar_251120' into update_fe_forcing

mnlevy1981 committed:
Updates to iron forcing from K Moore

mnlevy1981 committed:
Merge branch 'dev/ncar' into update_fe_forcing

mnlevy1981 committed:
Revert changes to add num_pass arguments

We decided to update the MARBL interface instead of playing with how MOM6
smooths the MARBL ICs, so we don't need to modify src/framework or
src/initialization in this PR.

Also changed some ".eq." to "==" in MARBL_tracers.F90

mnlevy1981 commented:
With https://github.com/ESCOMP/MOM_interface/pull/231 and https://github.com/ESCOMP/FMS/pull/7, this passes `aux_mom_MARBL` and `pr_mom` (the latter includes baseline comparisons to baselines generated with the head of `dev/ncar`); this is ready for final review / merging

alperaltuntas commented on src/tracer/MARBL_tracers.F90:
```diff
@@ -135,8 +135,9 @@ module MARBL_tracers
-  character(len=200) :: feventflux_file  !< name of [netCDF] file containing iron vent flux
+  character(len=200) :: fesedflux_file      !< name of [netCDF] file containing iron sediment flux
+  character(len=200) :: fesedfluxred_file   !< name of [netCDF] file containing reduced iron sediment flux
+  character(len=200) :: feventflux_file     !< name of [netCDF] file containing iron vent flux
```
Optional: While it's probably unlikely to hit the character limit of 200, I suggest using deferred-length character strings, which are more robust and memory-efficient.

alperaltuntas reviewed (commented):
Below is a minor suggestion. Aside from that, have you run the DIMCSL test? It tends to timeout because reading in JRA datastreams take so long. Unless you were able to run it successfully, I suggest running a DIMCSL test with CORE2 with and without MARBL.

mnlevy1981 commented:
@alperaltuntas -- I opened #410 to note that it would be nice to switch to deferred-length character strings, but I'd like to put that off until a future PR. I forgot to mention that the DIMCSL tests in `pr_mom` did not have enough walltime; I have resubmitted two tests (`G_JRA` and `G1850MARBL_JRA`) with an 8 hour walltime request, but they've been in the queue all day. Hopefully they'll run over the weekend and I can report back on Monday -- both are running on my branch, neither will have a baseline to compare to but the rest of the `pr_mom` tests show that non-MARBL compsets are not changing answers.

mnlevy1981 commented:
DIMCSL tests ran over the weekend; the good news is that the `G_JRA` test passed:

```
PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel COMPARE_dimscale_1_Tp_base
PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel COMPARE_dimscale_2_Lp_base
PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel COMPARE_dimscale_3_Hp_base
PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel COMPARE_dimscale_4_Zp_base
PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel COMPARE_dimscale_5_Rp_base
PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel COMPARE_dimscale_6_Qp_base
```

Unfortunately, there are a few issues with the MARBL compset:

```
FAIL DIMCSL_Ld1.TL319_t232.G1850MARBL_JRA.derecho_intel COMPARE_dimscale_1_Tp_base
FAIL DIMCSL_Ld1.TL319_t232.G1850MARBL_JRA.derecho_intel COMPARE_dimscale_2_Lp_base
FAIL DIMCSL_Ld1.TL319_t232.G1850MARBL_JRA.derecho_intel COMPARE_dimscale_3_Hp_base
PASS DIMCSL_Ld1.TL319_t232.G1850MARBL_JRA.derecho_intel COMPARE_dimscale_4_Zp_base
PASS DIMCSL_Ld1.TL319_t232.G1850MARBL_JRA.derecho_intel COMPARE_dimscale_5_Rp_base
PASS DIMCSL_Ld1.TL319_t232.G1850MARBL_JRA.derecho_intel COMPARE_dimscale_6_Qp_base
```

@alperaltuntas and @klindsay28 -- I think it's important that we fix this before the CESM3 release, but I'm okay creating a new issue ticket about this test failure and tackling it in a future PR if that's okay with the two of you. Otherwise I'll fix it in this PR.

mnlevy1981 commented:
I am currently running a DIMCSL test with MARBL out of the current `dev/ncar`; I added `DEBUG=TRUE` to `user_nl_mom` for the control case as well as the three failing tests. My current thinking is we should wait to see these test results -- if the failure is already on `dev/ncar`, we can fix it later (and I may have checksum output that helps find the problem(s)) but if it was introduced in this PR we should fix it here as well.

alperaltuntas commented:
Thanks @mnlevy1981. My experience is that CORE2 DIMCSL tests run much faster compared to JRA tests since datm doesn't have to slowly read through the relatively large JRA datastreams, but I understand if CORE2 testing is not meaningful in your case.

mnlevy1981 commented:
@alperaltuntas I've verified that the DIMCSL test is failing for MARBL compsets in `dev/ncar`, so it is not related to this PR. I'll open a separate issue ticket to discuss details of that failure, but I think this one is okay to bring in as-is and we'll fix the dimensional scaling in a future PR

alperaltuntas commented:
@mnlevy1981 Seems like there is at least one build failure in CI tests, but otherwise, LGTM.

mnlevy1981 committed:
Fix CI build error (hopefully)

CI was complaining about dust_ratio**-0.9 ("Unary operator following arithmetic
operator (use parentheses)") so I took the message's advice and changed to

dust_ratio**(-0.9)

mnlevy1981 commented:
> @mnlevy1981 Seems like there is at least one build failure in CI tests, but otherwise, LGTM.

Fixed! Sorry about that, the first few times I looked through the failed test log I had trouble finding the actual cause of the failure (gfortran wants `dust_ratio**(-0.9)` instead of `(dust_ratio**-0.9)`). Tests are passing now!

klindsay28 reviewed (approved):
LGTM
